codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AddAwsConfigFields.java

```diff
@@ -0,0 +1,148 @@
+        return SymbolUtils.createValueSymbolBuilder(symbolName)
+                .putProperty(SymbolUtils.GO_UNIVERSE_TYPE, true).build();
+    }
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            TriConsumer<String, String, Consumer<GoWriter>> writerFactory
+    ) {
+        LOGGER.info("generating aws.Config based client constructor");
+        writerFactory.accept("api_client.go", settings.getModuleName(), w -> {
+            writeAwsConfigConstructor(model, model.expectShape(settings.getService()).asServiceShape().get(), w);
+        });
+    }
+
+    @Override
+    public List<RuntimeClientPlugin> getClientPlugins() {
+        return ListUtils.of(
+                RuntimeClientPlugin.builder()
+                        .configFields(UNIVERSAL_FIELDS)
+                        .build(),
+                RuntimeClientPlugin.builder()
+                        .configFields(ListUtils.of(ConfigField.builder()
+                                .name(CREDENTIALS_CONFIG_NAME)
+                                .type(getAwsCoreSymbol("CredentialsProvider"))
+                                .documentation("The credentials object to use when signing requests.")
+                                .build()))
+                        .servicePredicate((model, serviceShape) -> model.getKnowledge(ServiceIndex.class)
+                                .getAuthSchemes(serviceShape).values().stream().anyMatch(trait -> trait.getClass()
+                                        .equals(SigV4Trait.class)))
+                        .build()
+        );
+    }
+
+    private void writeAwsConfigConstructor(Model model, ServiceShape service, GoWriter writer) {
+        writer.writeDocs("NewFromConfig returns a new client from the provided config.");
+        writer.openBlock("func NewFromConfig(cfg $T, optFns ... func(*Options)) *Client {", "}",
+                getAwsCoreSymbol("Config"), () -> {
+                    writer.openBlock("opts := Options{", "}", () -> {
+                        writer.write("$L: cfg.$L,", HTTP_CLIENT_CONFIG_NAME, HTTP_CLIENT_CONFIG_NAME);
+                        for (RuntimeClientPlugin plugin : getClientPlugins()) {
+                            if (!plugin.matchesService(model, service)) {
+                                continue;
+                            }
+                            plugin.getConfigFields().forEach(configField -> {
+                                writer.write("$L: cfg.$L,", configField.getName(), configField.getName());
+                            });
+                        }
+                    });
+                    writer.write("");
+                    writer.openBlock("for _, fn := range optFns {", "}", () -> {
+                        writer.write("fn(&opts)");
+                    });
+                    writer.write("return New(opts)");
```

Comment:
I think the generated API client generic constructor needs an update as well to not return `error`